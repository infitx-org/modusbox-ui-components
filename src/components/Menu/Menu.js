import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import * as utils from '../../utils/common';
import Icon from '../Icon';
import './Menu.scss';
import '../../icons/mule/back-small.svg';

const bindOnClickProp = onClick => element =>
  React.cloneElement(element, {
    ...element.props,
    onClick,
  });

const bindPathnameProp = pathname => element =>
  React.cloneElement(element, {
    ...element.props,
    pathname,
  });

const bindActiveProp = pathname => element => {
  const { path, back, active } = element.props;
  const matchesPath = path !== undefined && path === pathname && !back;
  if (matchesPath || active) {
    return React.cloneElement(element, {
      ...element.props,
      active: true,
    });
  }
  return element;
};
const bindKey = (element, index) =>
  React.cloneElement(element, {
    ...element.props,
    key: index.toString(),
  });

const bindDisabledProp = disabled => element =>
  React.cloneElement(element, {
    ...element.props,
    disabled: element.props.disabled || disabled,
  });

/* eslint-disable */
const isMenuSection = element => element.type === MenuSection;
const isMenuItem = element => element.type === MenuItem;
/* eslint-enable */

const wrapItemsInSections = items => {
  const groupedMenuItems = [];
  let currentGroup = [];
  const addCurrenntGroupToGroupedItems = () => {
    if (currentGroup.length) {
      groupedMenuItems.push(<MenuItemsGroup>{currentGroup}</MenuItemsGroup>);
      currentGroup = [];
    }
  };

  items.forEach(node => {
    if (isMenuSection(node) || node.props.back) {
      addCurrenntGroupToGroupedItems();
      groupedMenuItems.push(node);
    } else {
      currentGroup.push(node);
    }
  });

  addCurrenntGroupToGroupedItems();
  return groupedMenuItems.map(bindKey);
};

class MenuItem extends PureComponent {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    if (!this.props.disabled) {
      this.props.onClick(this.props.to || this.props.path);
    }
  }
  render() {
    const { label, disabled, hidden, active, back } = this.props;
    if (hidden) {
      return null;
    }
    let BackIcon = null;
    if (back) {
      BackIcon = (
        <Icon className="element-menu__item__back-icon" name="arrow" size={10} fill="#999" />
      );
    }
    const classNames = utils.composeClassNames([
      'element-menu__item',
      active && 'element-menu__item--active',
      disabled && 'element-menu__item--disabled',
      back && 'element-menu__item--back',
    ]);

    return (
      <div className={classNames} onClick={this.onClick} role="presentation">
        {BackIcon}
        {label}
      </div>
    );
  }
}

MenuItem.defaultProps = {
  path: undefined,
  to: undefined,
  disabled: false,
  hidden: false,
  back: false,
};
MenuItem.propTypes = {
  path: PropTypes.string,
  to: PropTypes.string,
  disabled: PropTypes.bool,
  hidden: PropTypes.bool,
  back: PropTypes.bool,
};

const MenuSection = ({ pathname, label, children, onClick, hidden, disabled }) => {
  if (hidden) {
    return null;
  }
  const menuItems = React.Children.toArray(children)
    .filter(element => isMenuItem(element))
    .map(bindOnClickProp(onClick))
    .map(bindActiveProp(pathname))
    .map(bindDisabledProp(disabled));

  let menuSectionLabel = null;
  if (label) {
    menuSectionLabel = <div className="element-menu__section-label">{label}</div>;
  }
  return (
    <div className="element-menu__section">
      {menuSectionLabel}
      <div className="element-menu__section-items">{menuItems}</div>
    </div>
  );
};

MenuSection.defaultProps = {
  label: undefined,
};
MenuSection.propTypes = {
  label: PropTypes.string,
};

const MenuItemsGroup = ({ children }) => (
  <div className="element-menu__section-items">{children}</div>
);

class Menu extends PureComponent {
  static flattenMenuSections(items) {
    // It flattens nested MenuSection, MenuItem components into a one-level array
    return items.reduce(
      (prevItems, currentItem) => [
        ...prevItems,
        ...(isMenuSection(currentItem)
          ? Menu.flattenMenuSections(React.Children.toArray(currentItem.props.children))
          : [currentItem]),
      ],
      []
    );
  }
  constructor() {
    super();
    this.getActiveRoot = this.getActiveRoot.bind(this);
  }
  getActiveRoot(parentNode) {
    const { pathname } = this.props;
    let activeRoot = null;
    // render Menu when pathname matches path
    if (parentNode.props.path !== undefined && parentNode.props.path === pathname) {
      return parentNode;
    }
    // Default to Menu when going manual - no route matching
    if (parentNode === this && parentNode.props.pathname === undefined) {
      activeRoot = parentNode;
    }

    // Flatten MenuSections in order not to have nested children when detecting active menu
    const items = Menu.flattenMenuSections(React.Children.toArray(parentNode.props.children));

    items.some(node => {
      // find the first matching menu item and return the parent or the item itself
      // depending if needs to be treated like a root
      const { path, asRoot, children, active } = node.props;
      const hasChildren = children !== undefined;
      const pathMatches = path !== undefined && path === pathname;
      if (isMenuItem(node)) {
        if (pathMatches || active) {
          // asRoot prop is meant to be used when menu has child elements
          // and we do not want to render the parent node but the child nodes
          activeRoot = asRoot ? node : parentNode;
        } else if (hasChildren) {
          activeRoot = this.getActiveRoot(node);
        }
      }
      return activeRoot;
    });

    return activeRoot;
  }

  render() {
    const { pathname, onChange } = this.props;
    let menuComponents = null;
    const activeRoot = this.getActiveRoot(this);
    if (activeRoot !== null) {
      menuComponents = React.Children.toArray(activeRoot.props.children);
      menuComponents = wrapItemsInSections(
        menuComponents
          .filter(element => isMenuItem(element) || isMenuSection(element))
          .map(bindOnClickProp(onChange))
          .map(bindPathnameProp(pathname))
          .map(bindActiveProp(pathname))
      );
    }
    return <div className="mb-element element-menu">{menuComponents}</div>;
  }
}

Menu.defaultProps = {
  path: undefined,
  pathname: undefined,
};
Menu.propTypes = {
  // eslint-disable-next-line react/no-unused-prop-types
  path: PropTypes.string,
  pathname: PropTypes.string,
  onChange: PropTypes.func.isRequired,
};

export default Menu;
export { MenuItem, MenuSection };
