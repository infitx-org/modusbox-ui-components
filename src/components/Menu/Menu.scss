@import '../../assets/styles/vars/colors.scss';

.element-menu {
  overflow: hidden;
  background: $color__main-disabled;
}
.element-menu__section {
}
.element-menu__section-label {
  text-align: center;
  width: 100%;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: left;
  padding-left: 10px;
  background: #fff;
}
.element-menu__section-items{
  margin-top: 1px;
  margin-bottom: 1px;
  background: #fafafa;
}

.element-menu__item {
  display: flex;
  align-items: center;
  width: 100%;
  padding-left: 16px;
  height: 30px;
  cursor: pointer;
  line-height: 30px;
  border-left: 4px solid transparent;
  transition: border 0.2s linear, color 0.2s linear, fill 0.2s linear;
  font-size: 12px;

  &:hover:not(.element-menu__item--active) {
    color: $color__main-primary;
  }
}

.element-menu__item--active {
  background: #F2F1F1;
  font-weight: bold;
  color: #333;
}
.element-menu__item--disabled {
  color: #999;
  fill: #ccc;
  border-color: transparent;
}
.element-menu__item--back {
  padding-left: 0px;
  height: 40px;
  background: #fff;
  margin-top: 1px;
}

.element-menu__item__back-icon {
  transform: rotateZ(180deg);  
  margin-left: 3px;
  margin-right: 3px;
}
