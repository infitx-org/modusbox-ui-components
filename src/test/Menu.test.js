import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Menu, { MenuItem, MenuSection } from '../components/Menu';

const onChangeMockEvent = jest.fn();

it('renders only the menu', () => {
  const wrapper = mount(<Menu path="/" pathname="/" onChange={onChangeMockEvent} />);
  expect(wrapper.find('.element-menu')).toHaveLength(1);
});

it('renders the menu wrapper also if pathname does not match any route', () => {
  const wrapper = mount(<Menu path="/" pathname="/test" onChange={onChangeMockEvent} />);
  expect(wrapper.find('.element-menu')).toHaveLength(1);
});

it('renders the menu items if parent root matches', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/" onChange={onChangeMockEvent}>
      <MenuItem path="/foo" label="foo" />
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  expect(wrapper.find('.element-menu__item')).toHaveLength(2);
});

it('renders the menu items if child route matches', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <MenuItem path="/foo" label="foo" />
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  expect(wrapper.find('.element-menu__item')).toHaveLength(2);
});

it('does not render the menu items if no route matches', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/non-existing" onChange={onChangeMockEvent}>
      <MenuItem path="/foo" label="foo" />
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  expect(wrapper.find('.element-menu__item')).toHaveLength(0);
});

it('renders the nested child when asRoot prop is set and route matches parent', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <MenuItem path="/foo" label="foo" asRoot>
        <MenuItem path="/foo/nested" label="nested" />
      </MenuItem>
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  const menuItem = wrapper.find('.element-menu__item');
  expect(menuItem).toHaveLength(1);
  expect(menuItem.text()).toBe('nested');
  expect(wrapper.find('.element-menu__item')).toHaveLength(1);
});

it('renders the parent child if asRoot prop is not set and route matches parent', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <MenuItem path="/foo" label="foo">
        <MenuItem path="/foo/nested" label="nested" />
      </MenuItem>
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  const menuItems = wrapper.find('.element-menu__item');
  expect(menuItems).toHaveLength(2);
  expect(wrapper.find('.element-menu__item')).toHaveLength(2);
  expect(menuItems.at(0).text()).toBe('foo');
  expect(menuItems.at(1).text()).toBe('bar');
});

it('renders the nested child if matching if asRoot prop is not set ', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/1st/2nd/3rd/4th" onChange={onChangeMockEvent}>
      <MenuItem path="/1st" label="1st Level">
        <MenuItem path="/1st/2nd" label="2st level">
          <MenuItem path="/1st/2nd/3rd" label="3nd level">
            <MenuItem path="/1st/2nd/3rd/4th" label="4th level" />
          </MenuItem>
        </MenuItem>
      </MenuItem>
    </Menu>,
  );

  const menuItem = wrapper.find('.element-menu__item');
  expect(menuItem).toHaveLength(1);
  expect(menuItem.text()).toBe('4th level');
});

it('renders the menu items when partial match is allowed', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/foo/partial/match" onChange={onChangeMockEvent}>
      <MenuItem path="/foo" label="foo" partial />
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  const menuItem = wrapper.find('.element-menu__item');
  const activeMenuItem = wrapper.find('.element-menu__item--active');
  expect(menuItem).toHaveLength(2);  
  expect(activeMenuItem.text()).toBe('foo');
});

it('renders the menu section if parent root matches', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/" onChange={onChangeMockEvent}>
      <MenuSection label="Menu Section">
        <MenuItem path="/foo" label="foo" />
        <MenuItem path="/bar" label="bar" />
      </MenuSection>
    </Menu>,
  );
  const menuSection = wrapper.find('.element-menu__section');
  const menuSectionLabel = menuSection.find('.element-menu__section-label');
  expect(menuSectionLabel).toHaveLength(1);
  expect(menuSectionLabel.text()).toBe('Menu Section');
});

it('renders the Menu when no pathname is set', () => {
  const wrapper = mount(
    <Menu onChange={onChangeMockEvent}>
      <MenuItem label="foo" />
      <MenuItem label="bar" />
    </Menu>,
  );
  const menuItems = wrapper.find('.element-menu__item');
  expect(menuItems).toHaveLength(2);
});

it('renders the manually set active prop', () => {
  const wrapper = mount(
    <Menu onChange={onChangeMockEvent}>
      <MenuItem label="foo" active />
      <MenuItem label="bar" />
    </Menu>,
  );
  const activeMenuItem = wrapper.find('.element-menu__item--active');
  expect(activeMenuItem).toHaveLength(1);
  expect(activeMenuItem.text()).toBe('foo');
});

it('renders the disabled prop on menu item', () => {
  const wrapper = mount(
    <Menu onChange={onChangeMockEvent}>
      <MenuItem label="foo" disabled />
      <MenuItem label="bar" />
    </Menu>,
  );
  const disabledMenuItem = wrapper.find('.element-menu__item--disabled');
  expect(disabledMenuItem).toHaveLength(1);
  expect(disabledMenuItem.text()).toBe('foo');
});

it('does not render a hidden menu item', () => {
  const wrapper = mount(
    <Menu onChange={onChangeMockEvent}>
      <MenuItem label="foo" hidden />
      <MenuItem label="bar" />
    </Menu>,
  );
  const menuItems = wrapper.find('.element-menu__item');
  expect(menuItems).toHaveLength(1);
});

it('does not render a hidden menu section', () => {
  const wrapper = mount(
    <Menu onChange={onChangeMockEvent}>
      <MenuSection hidden>
        <MenuItem label="foo" />
        <MenuItem label="bar" />
      </MenuSection>
    </Menu>,
  );
  const menuSection = wrapper.find('.element-menu__section');
  expect(menuSection).toHaveLength(0);
});

it('trigger onChange when clicking a menu item', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(
    <Menu path="/" pathname="/" onChange={mockEvent}>
      <MenuItem path="/foo" label="foo" />
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  const menuItems = wrapper.find('.element-menu__item');
  menuItems.at(0).simulate('click');
  expect(mockEvent).toHaveBeenCalled();
});

it('trigger onChange with correct value when clicking a menu item', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(
    <Menu path="/" pathname="/" onChange={mockEvent}>
      <MenuItem path="/foo" label="foo" />
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  const menuItems = wrapper.find('.element-menu__item');
  menuItems.at(1).simulate('click');
  expect(mockEvent).toHaveBeenCalledWith('/bar');
});

// Snapshot
it('renders the menu correctly when multiple props are set', () => {
  const wrapper = mount(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <MenuItem path="/foo" label="foo" asRoot>
        <MenuItem path="/foo/nested" label="nested" />
      </MenuItem>
      <MenuItem path="/bar" label="bar" />
    </Menu>,
  );
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
