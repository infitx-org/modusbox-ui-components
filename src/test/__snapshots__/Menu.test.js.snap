// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the menu correctly when multiple props are set 1`] = `
<Menu
  onChange={[MockFunction]}
  path="/"
  pathname="/foo"
>
  <div
    className="mb-element element-menu"
  >
    <MenuItemsGroup
      key="0"
    >
      <div
        className="element-menu__section-items"
      >
        <MenuItem
          asRoot={false}
          back={false}
          disabled={false}
          hidden={false}
          key=".0"
          label="nested"
          onClick={[MockFunction]}
          partial={false}
          path="/foo/nested"
          pathname="/foo"
          size={10}
        >
          <div
            className="element-menu__item"
            onClick={[Function]}
            role="presentation"
          >
            nested
          </div>
        </MenuItem>
      </div>
    </MenuItemsGroup>
  </div>
</Menu>
`;
